### ISSUE TITLE
Pre-fund account success, then add nodes failed, No pre-fund account, add nodes successfully

### ISSUE THREAD

**Issue by ljbb** [ethereum/go-ethereum#15354]
#### System information

Geth version: `1.7.2`
OS & Version: Linux
Commit hash : (`develop`)

#### Expected behaviour
After pre-fund account in private-network successfully, add nodes by using admin.addPeer("publickey@ip:port") function could find other node with the same chainId and newtworkid.

#### Actual behaviour
After pre-fund account in private-network successfully, add nodes by using admin.addPeer("publickey@ip:port") function never add peer successfully.

But, without pre-fund, addPeer function works as always.

#### Steps to reproduce the behaviour
1. Generate new account before init the genesis block.
[CMD] geth --datadir "/path" --password "/path/passwd.txt" account new
2. vi genesis.json and put the new account into the file with specific number of ether value.
3. generate the genesis block.
[CMD] geth --datadir "/path" init /path/genesis.json
4. launch the console.
[CMD] geth --rpc --rpccorsdomain "*" --datadir "/path" --port "30303" --rpcapi "db,eth,net,web3" --networkid xxxxxx console
5. run up another node without step 1,2
6. use admin.addPeer("publickey@ip:port") function to add the node, system return value: true
7. admin.peers shows nothing.

If skip the Pre-fund step, the add peer action is successful.

Confirm no network issue.

#### genesis.json content
{       

        "config": {
                "chainId": xxxxxx,
                "homesteadBlock": 0,
                "eip155Block": 0,
                "eip158Block": 0
        },
	"nonce": "0x00",
	"timestamp": "0x00",
	"parentHash": "0x0000000000000000000000000000000000000000000000000000000000000000",
	"extraData": "0x00",
	"gasLimit": "0x2fefd8",
	"difficulty": "0x20000",
	"mixhash": "0x0000000000000000000000000000000000000000000000000000000000000000",
	"coinbase": "0x0000000000000000000000000000000000000000",
	"alloc": {
                "0x4987f42b3ba00f90a3a02f5f34588bc8386ec495":{"balance":"999000000000000000000000000"}
        }
}

**Comment by karalabe** [comment]
I'm betting your `geth --datadir "/path" init /path/genesis.json` is failing with an error that you already inited your chain to an old genesis, which you just ignore, and thus the two nodes get different genesis blocks.

**Comment by ljbb** [comment]
hmm... those two nodes already used the same chainId as well as network id, so how to identify my two nodes get different genesis?

**Comment by manicprogrammer** [comment]
**SO...**
I wrote all of the below and then I remembered you said it worked fine if you didn't pre-fund the genesis block. In which case I don't think any of the below helps out as it works for you otherwise. I used the same genesis.json you posted obviousuly with a unique chainId and appropriate account and got the below results. Sorry it is not likely to help especially after I wrote all that. tl;dr

Since I wrote all of the below I'll leave it in case someone else has a similar issue but is helped by the below.
- - - - - -

I followed the above steps. In the first case it failed and in the second it performed as expected. The difference was the IP address I utilized to addPeer. For node 1 getting the nodeInfo gave me the following:

> admin.nodeInfo.enode
> "enode://768ab38ebbea826e27c0b6e68f3d2c15563a18e207f9eef1eff1b70b67da86cbb42c8c15531e331597dee61157b360cfadc130399b1d7b6f85b446b1d451584a@72.xx.xx.xx:30308"
(note the public ip)

If I used the enode URI with the public IP for me it never added the node. I can feel pretty certain due to my network sitting behind a router in my home and how I certainly don't have the ports I was testing with in this case open on the router and directing to that machine. I'll try it out with the ports open as well and see what happens. Also note I did not provide node 1 with that public IP. The machine it is on only has a local network IP- that public IP is the router ip that the geth software is picking up on its own.

If instead I put the local network ip into the URI instead of the public IP and called addPeer it worked almost always. When it didn't work I was getting an I/O timeout this is likely due to me logging heavy to a non-SSD drive for the test and the machine was doing a lot of other things. It certainly appeared to be more the environment of the process than the process itself. I did note that sometimes after calling addPeer it could be a small handful of seconds before I would see it appear. If I checked instantly after calling it was not uncommon to not be showing up. If I kept checking after a short time it did appear. 

I noticed the lag often seemed to correlate with if it had found another peer to talk to and determining if it should be added. I'd see a remote machine in the peers for a bit and then it would be gone.  Sometimes I'd see a remote machine in there be gone another instantly appear then it disappeared and then my addPeer entry would show up. I apologize I don't know (yet) how it threads those operations out. It certainly looked like there could be a little queueing going on  (just my conjecture based on observation) regardless for that environment it was running it seemed slower to show up if it was already evaluating for another peer. These other nodes appeared temporarily despite having a non-standard networkId and chainId. Perhaps there was collision in my choices or just a slow evaluation process that I would sometimes catch.

Bottom line- based on the above and that I also I researched the code related to the `geth account new` @karalabe thought that it was related to an issue with the genesis block and non-ideal operation of the account commands (see #15374) but that doesn't seem to hold up after some investigation, if I am understanding correctly.  

I suspect it's one or combination of these that is really environmental:

- use of an ip (public ip) that doesn't route properly or have the needed ports open
- need to give it a handful of seconds to see if the peer gets added as there can be some lag
- system is too busy and you get an I/O timeout

**Comment by manicprogrammer** [comment]
I am going to dig in deeper into this. Apparently there are a number of similar reports that are expected to be related. There has to be a findable cause.
